pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -308,3 +306,22 @@ def is_module(self) -> bool:
         from pytorch_lightning import LightningDataModule, LightningModule  # prevent cyclic import
 
         return isinstance(self.instance, (LightningModule, LightningDataModule))
+
+
+@dataclass
+class _DataHookSource:
+    """Stores the info about the common DataHooks within LightningModule and LightningDataModule.
+
+    The hook source can be
+
+    1. a method from the :class:`~pytorch_lightning.core.lightning.LightningModule`,
+    2. a method from the :class:`~pytorch_lightning.core.datamodule.LightningDataModule`,
+
+    Arguments:
+        instance: A LightningModule or a LightningDataModule
+    """
+
+    instance: Optional[Union["pl.LightningModule", "pl.LightningDataModule"]]
+
+    def get_hook(self, hook_name):
+        return getattr(self.instance, hook_name)
```

Comment:
setup and teardown are also part of DataHooks: https://github.com/PyTorchLightning/pytorch-lightning/blob/2c7c4aab8087d4c1c99c57c7acc66ef9a8e815d4/pytorch_lightning/core/hooks.py#L373-L399

This could easily be misused to call `setup` or `teardown` only once between DataModule and LightningModule instead of on both

Reply:
but these are not specific to just data. For eg one can initialize a module component inside `LightningModule.setup` and some data-specific component inside `Datamodule.setup`, so we need to call both in such a case. A similar case can happen for teardown.